File: leet-code/496.下一个更大元素-i.js
/*
 * @lc app=leetcode.cn id=496 lang=javascript
 *
 * [496] 下一个更大元素 I
 */

// @lc code=start
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function (nums1, nums2) {
  const numMap = new Map()
  const res = Array(nums1.length).fill(-1)
  nums1.forEach((num, i) => {
    numMap.set(num, i)
  })

  // 单调栈
  const stack = []
  for (let i = 0; i < nums2.length; i++) {
    const item = nums2[i];
    while (stack.length) {
      const top = stack.pop()
      if (item > top) {
        if (numMap.has(top)) {
          res[numMap.get(top)] = item
        }
      } else {
        stack.push(top)
        break
      }
    }
    stack.push(item)
  }
  return res
};
// @lc code=end

File: leet-code-rs/src/solutions/496_下一个更大元素-i.rs
/*
 * @lc app=leetcode.cn id=496 lang=rust
 *
 * [496] 下一个更大元素 I
 */

// @lc code=start

use std::collections::HashMap;

#[allow(unused)]
impl Solution {
    pub fn next_greater_element(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let mut res = vec![-1; nums1.len()];

        let map: HashMap<i32, usize> = nums1
            .iter()
            .enumerate()
            .map(|(idx, &value)| (value, idx))
            .collect();
        let mut stack = vec![];

        for n in nums2 {
            while !stack.is_empty() && stack.last() < Some(&n) {
                let n2 = stack.pop().unwrap();
                if map.contains_key(&n2) {
                    let idx = map.get(&n2).unwrap();
                    res[*idx] = n;
                }
            }
            stack.push(n);
        }

        return res;
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_next_greater_element() {
        //println!("res is {:?}", Solution::next_greater_element(vec![]));
        //assert_eq!(Solution::next_greater_element(vec![]),[]);
        assert!(true)
    }
}
